PR related to issue #217.

issactoast opened the pull request:
Hi @juliasilge 

This PR related to issue #217 

I have been thinking about this since I found `strat_sample` sounds good to me, so I fork the package and let my data flow the functions. 

I found two parts is switched which make the result different than `sample_frac` in `dplyr`

I twicked `mc_splits` input from 1-prop to prop, so that the functions in mc.R find the index for prop. And I finally use `ind` for the analysis. 

These two chance makes the result from rsample and dplyr to be the same as follows;

``` r
#sample_frac and sample_n
library(dplyr)

set.seed(2021)
dim(airquality)
#> [1] 153   6

#sample 10% from each month
airquality_frac <- airquality %>% 
  group_by(Month) %>%  
  sample_frac(size = 0.1, replace=FALSE)

library(rsample)

airquality_split <- initial_split(airquality, prop = 0.1,
                                  strata = "Month")

airquality_frac2 <- training(airquality_split)

dim(airquality_frac)
#> [1] 15  6
dim(airquality_frac2)
#> [1] 15  6
```

Could you give me a feedback? I have feeling that this resolves the issue (user expectation about prop input) and increases the coherence between the tidymodels and the tidyverse packages. What do you think?

issactoast commented:
I resolved the testthat package issue and the prop change in validation_split.R

DavisVaughan commented on R/mc.R:
```diff
@@ -61,7 +61,7 @@ mc_cv <- function(data, prop = 3/4, times = 25, strata = NULL, breaks = 4, ...)
   split_objs <-
     mc_splits(data = data,
-              prop = 1 - prop,
+              prop = prop, # prop for train set
```
Could you please remove the comment? We can always look back at this PR and the associated issue if we are wondering why the code is written this way

DavisVaughan commented on R/mc.R:
```diff
@@ -83,8 +83,8 @@ mc_cv <- function(data, prop = 3/4, times = 25, strata = NULL, breaks = 4, ...)
 mc_complement <- function(ind, n) {
-  list(analysis = setdiff(1:n, ind),
-       assessment = ind)
+  list(analysis = ind, # ind for analysis
```
Same, no comment needed

DavisVaughan commented on R/mc.R:
```diff
@@ -83,8 +83,8 @@ mc_cv <- function(data, prop = 3/4, times = 25, strata = NULL, breaks = 4, ...)
 
 # Get the indices of the analysis set from the assessment set
```
Could you please update this comment to say:

```
# Get the indices of the assessment set from the analysis set
```

DavisVaughan commented on R/validation_split.R:
```diff
@@ -35,7 +35,7 @@ validation_split <- function(data, prop = 3/4, strata = NULL, breaks = 4, ...) {
   split_objs <-
     mc_splits(data = data,
-              prop = 1 - prop,
+              prop = prop, # prop for analysis
```
Same, no comment required

DavisVaughan commented on tests/testthat/test_validation.R:
```diff
@@ -49,8 +49,8 @@ test_that('strata', {
-  expect_true(all(sizes3$analysis == 42))
-  expect_true(all(sizes3$assessment == 12))
+  expect_true(all(sizes3$analysis == 39))
+  expect_true(all(sizes3$assessment == 15))
```
Could you please also the following test to `test_initial.R`:

```r
test_that("`prop` computes the proportion for analysis (#217)", {
  set.seed(11)
  
  props <- c(.1, .9)
  
  for (prop in props) {
    # Not stratified
    split <- initial_split(airquality, prop = prop)
    actual <- nrow(analysis(split))
    expect <- as.integer(floor(nrow(airquality) * prop))
    
    expect_identical(actual, expect)
    
    # Stratified
    split <- initial_split(airquality, prop = prop, strata = Month)
    actual <- nrow(analysis(split))
    expect <- as.integer(sum(floor(table(airquality$Month) * prop)) )
    
    expect_identical(actual, expect)
  }
})
```

This should ensure that this can't happen again. Adding a test like this is also a more robust alternative for the comments you had added directly in the code (which I requested you remove)

DavisVaughan reviewed (changes requested):
Thanks @issactoast! This looks like the right thing to do (as noted in https://github.com/tidymodels/rsample/issues/217#issuecomment-782151875). I have a few comments. Once you make these changes, I'll merge.

Notes to self:
- Rebase on master before merging
- Add a NEWS bullet (wait until after rebasing to keep this news entry separate from the 0.0.9 ones)

issactoast commented:
I have 
- removed and updated the comments on mc.R and  validation_split.R.
- added testthat code to the test_initial.R

Wow, this PR is my first contribution to the R community! It feels awesome! Thank you @DavisVaughan and @juliasilge! :)

DavisVaughan commented:
Perfect, I'll take it from here!

DavisVaughan commented:
Thanks @issactoast! That was a very useful first contribution!

issactoast commented:
@DavisVaughan Wonderful! Thank you for the concrete comment! 👍
